knowledge-bases/advance-concepts/0_chunk_size_evaluation_for_KB_RAG.ipynb

```diff
@@ -0,0 +1,1844 @@
+    "- Create a data source within knowledge base which will connect to Amazon S3\n",
+    "- Once the data is available in the Bedrock Knowledge Bases with different chunk size, we'll evaluate the text chunks retreived for refernce QA pairs from these knowledge bases for faithfulness, correctness, and relevancy metrics using LlamaIndex. \n",
+    "- Using these metrics we can decide for RIGHT chunk size for our RAG based application. \n",
+    "\n",
+    "Finally, based on the evaluation results, a question answering application with RIGHT chunk strategy can be built using the Amazon Bedrock APIs.  "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Usecase\n",
+    "### Dataset\n",
+    "In this example, you will use several years of Amazon's Letter to Shareholders as a text corpus to perform Q&A on."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Setup \n",
+    "Before running the rest of this notebook, you'll need to run the cells below to (ensure necessary libraries are installed and) connect to Bedrock.\n",
+    "\n",
+    "<div class=\"alert alert-block alert-warning\">\n",
+    "<b>Warning:</b> Please ignore error messages related to pip's dependency resolver.\n",
+    "</div>"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "%pip install --upgrade pip\n",
+    "%pip install boto3==1.33.2 --force-reinstall --quiet\n",
+    "%pip install botocore==1.33.2 --force-reinstall --quiet\n",
+    "%pip install -U opensearch-py==2.3.1\n",
+    "%pip install -U retrying==1.3.4\n",
+    "\n",
+    "%pip install langchain==0.0.342 --force-reinstall --quiet\n",
+    "%pip install llama-index==0.9.3.post1 --force-reinstall --quiet"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "collapsed": false,
+    "jupyter": {
+     "outputs_hidden": false
+    }
+   },
+   "outputs": [],
+   "source": [
+    "# restart kernel\n",
+    "from IPython.core.display import HTML\n",
+    "HTML(\"<script>Jupyter.notebook.kernel.restart()</script>\")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import warnings\n",
+    "warnings.filterwarnings('ignore')"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import nest_asyncio\n",
+    "nest_asyncio.apply()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import json\n",
+    "import os\n",
+    "import time\n",
+    "import boto3\n",
+    "import pandas as pd\n",
+    "from utility import create_bedrock_execution_role, create_oss_policy_attach_bedrock_execution_role, create_policies_in_oss\n",
```

Comment:
The utility.py file is outside the folder `advanced-concepts`, therefore, this might give error. Please make sure that it will run with `utility.py` being outside of the folder.